platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java

```diff
@@ -95,65 +100,59 @@ public boolean conflictExists() {
                         return true;
                     }
                 };
-                conflicts.add(new CapabilityConflict(group, name, candidatesForConflict));
-                return conflict;
             }
         }
         return PotentialConflictFactory.noConflict();
     }
 
-    private Set<NodeState> findNodesFor(CapabilityInternal capability) {
-        String capabilityId = capability.getCapabilityId();
-
-        return capabilityWithoutVersionToNodes.computeIfAbsent(capabilityId, k -> new LinkedHashSet<>());
-    }
-
     @Override
     public boolean hasConflicts() {
         return !conflicts.isEmpty();
     }
 
     @Override
     public void resolveNextConflict(Action<ConflictResolutionResult> resolutionAction) {
-        CapabilityConflict conflict = conflicts.poll();
+        String capabilityInConflict = conflicts.remove();
+        CapabilityConflict conflict = capabilityIdToConflict.remove(capabilityInConflict).withSelectedNodes();
+        capabilityWithoutVersionToNodes.put(capabilityInConflict, conflict.nodes);
```

Comment:
I wonder if we still need this line. now that we are filtering in `registerCandidate`

Reply:
I believe we do. While we filter at registration time, a node could be deselected between the registration and the conflict resolution itself.